# SEULTA pipeline

SEULTA: Simple Ensemble of Unsupervised Learners for Taxonomic Assignment

The SEULTA pipeline is shown in the below figure. <br />
![SEULTA testing](https://github.com/user-attachments/assets/a19b69bf-afd7-400c-bb2c-0a4af1b941e8) <br />

First, we split the input file into the sequences and the labels. From the sequences, we calculate the 7-mer count, 
which is reduced to two dimensions with t-SNE (using a PCA initiation, and Manhattan distance as the metric), and the 
normalized 7-mer counts, which are used to calculate silhouette scores. Using the dimensionally reduced 7-mer counts, 
we train n homogenous, concurrent ensembles (thus, bagging) of both K-Means (initiated with K-Means++, and only 
allowed to run for a single iteration) and K-Medoids (initiated with K-Medoids++, with a Manhattan distance metric,
also only allowed to run for a single iteration), with m voters. We choose one of the n ensembles for each model type, 
using the labels generated by the model, and the normalized 7-mer counts, by picking the model with the highest silhouette 
coefficient. Note that we use the normalized 7-mer counts, rather than the dimensionally reduced 7-mer counts, as otherwise 
we will end up with a pipeline that prefers models which cluster best in the dimensionally reduced axes, with no 
consideration of the original feature space (which is especially important, as t-SNE only preserves neighbourhoods, 
not distances; by calculating the silhouette coefficient using the non-dimensionally-reduced data, we can minimize 
how much the pipeline “overfits” (where by “overfits”, we mean a concept similar to that for classification) for the 
dimensionally-reduced data, and minimize how much it would fail to generalize to the original data). We then choose 
the best of the best two ensembles (where one is a bag of m K-Means, and the other a bag of m K-Medoids), using the 
silhouette coefficient, in the same manner as before. In order to be able to generate a clustering accuracy, we now 
map the model labels to the actual labels by use of the Hungarian Algorithm; that is, we find the optimal assignment 
between the clusters the models found and the true classes, such that the highest number of points are correctly 
assigned. We then return the mapped labels, the silhouette coefficient (found between the cluster labels and the 
normalized 7-mer counts), and the clustering accuracy. <br />

SEULTA outperforms simple learners, and closely approaches the performance of complex learner ensembles (DeLUCS-F, the SOTA), as shown in the figure below. K-means is used as the baseline due to its common use in this field [11-16]. <br />

![image](https://github.com/user-attachments/assets/64a57afe-056b-40ea-be67-6d4c1073e22e) <br />


# References
[1] C. C. Thompson, L. Chimetto, R. A. Edwards, J. Swings, E. Stackebrandt, and F. L. Thompson, “Microbial genomic taxonomy,” BMC Genomics, vol. 14, no. 1, p. 913, 2013, doi: 10.1186/1471-2164-14-913. <br />
[2] G. S. Randhawa, M. P. Soltysiak, H. E. Roz, C. P. de Souza, K. A. Hill, and L. Kari. Machine learning using intrinsic genomic signatures for rapid classification of novel pathogens: COVID-19 case study. PLoS ONE, 15(4):e0232391, 2020. <br />
[3] D. Emerson, L. Agulto, H. Liu, and L. Liu. Identifying characterizing bacteria in an era of genomics proteomics. BioScience, 58(10):925–936, 2008. <br />
[4] J. Avila Cartes, S. Anand, S. Ciccolella, P. Bonizzoni, and G. D. Vedova. Accurate fast clade assignment via deep learning frequency chaos game representation. Oxford University Press GigaScience, 12:giac119, 2022. <br />
[5] J. Rajkumari, P. Katiyar, S. Dheeman, P. Pandey, and D. K. Maheshwari, “The changing paradigm of rhizobial taxonomy and its systematic growth upto postgenomic technologies,” World J Microbiol Biotechnol, vol. 38, no. 11, p. 206, Nov. 2022, doi: 10.1007/s11274-022-03370-w. <br />
[6]	P. M. Arias, F. Alipour, K. A. Hill, and L. Kari, “DeLUCS: Deep learning for unsupervised clustering of DNA sequences,” PLOS ONE, vol. 17, no. 1, p. e0261531, Jan. 2022, doi: 10.1371/journal.pone.0261531. <br />
[7]	W. L. Applequist, “A brief review of recent controversies in the taxonomy and nomenclature of sambucus nigra sensu lato,” Acta Hortic., no. 1061, pp. 25–33, Jan. 2015, doi: 10.17660/ActaHortic.2015.1061.1. <br />
[8]	P. Karadayı Ataş, “Exploring the Molecular Interaction of PCOS and Endometrial Carcinoma through Novel Hyperparameter-Optimized Ensemble Clustering Approaches,” Mathematics, vol. 12, no. 2, p. 295, Jan. 2024, doi: 10.3390/math12020295. <br />
[9] D. Lewis, “Deep Learning of Unsupervised Clusters of DNA Sequences using Sequence Fragments,” Honours, University of Prince Edward Island, 2024. Accessed: Feb. 11, 2025. [Online]. Available: https://islandscholar.ca/islandora/object/ir%3A26198 <br />
[10] P. Millan Arias, K. A. Hill, and L. Kari, “iDeLUCS: a deep learning interactive tool for alignment-free clustering of DNA sequences,” Bioinformatics, vol. 39, no. 9, p. btad508, Sep. 2023, doi: 10.1093/bioinformatics/btad508. <br />
[11] F. Alipour, K. A. Hill, and L. Kari, “CGRclust: Chaos Game Representation for twin contrastive clustering of unlabelled DNA sequences,” BMC Genomics, vol. 25, no. 1, p. 1214, Dec. 2024, doi: 10.1186/s12864-024-11135-y. <br />
[12]	B. Yang et al., “MetaCluster: unsupervised binning of environmental genomic fragments and taxonomic annotation,” in Proceedings of the First ACM International Conference on Bioinformatics and Computational Biology, Niagara Falls New York: ACM, Aug. 2010, pp. 170–179. doi: 10.1145/1854776.1854803. <br />
[13]	J. Bao, R. Yuan, and Z. Bao, “An improved alignment-free model for dna sequence similarity metric,” BMC Bioinformatics, vol. 15, no. 1, p. 321, Sep. 2014, doi: 10.1186/1471-2105-15-321. <br />
[14]	A. Bustamam, H. Tasman, N. Yuniarti, Frisca, and I. Mursidah, “Application of k-means clustering algorithm in grouping the DNA sequences of hepatitis B virus (HBV),” presented at the INTERNATIONAL SYMPOSIUM ON CURRENT PROGRESS IN MATHEMATICS AND SCIENCES 2016 (ISCPMS 2016): Proceedings of the 2nd International Symposium on Current Progress in Mathematics and Sciences 2016, Depok, Jawa Barat, Indonesia, 2017, p. 030134. doi: 10.1063/1.4991238. <br />
[15]	G. Mendizabal-Ruiz, I. Román-Godínez, S. Torres-Ramos, R. A. Salido-Ruiz, H. Vélez-Pérez, and J. A. Morales, “Genomic signal processing for DNA sequence clustering,” PeerJ, vol. 6, p. e4264, Jan. 2018, doi: 10.7717/peerj.4264. <br />
[16]	I. Bonet, A. Escobar, A. Mesa-Múnera, and J. F. Alzate, “Consensus Clustering for Binning Metagenome Sequences,” in Advances in Soft Computing, O. Pichardo-Lagunas and S. Miranda-Jiménez, Eds., Cham: Springer International Publishing, 2017, pp. 273–284. doi: 10.1007/978-3-319-62428-0_23. <br />
[17]	R. Liao, R. Zhang, J. Guan, and S. Zhou, “A New Unsupervised Binning Approach for Metagenomic Sequences Based on N-grams and Automatic Feature Weighting,” IEEE/ACM Transactions on Computational Biology and Bioinformatics, vol. 11, no. 1, pp. 42–54, Jan. 2014, doi: 10.1109/TCBB.2013.137. <br />
